### PR TITLE
Initial support of or_patterns

### LINKED ISSUE #3454: Support or-patterns in let and if/while let expressions
The corresponding RFC: https://github.com/rust-lang/rfcs/pull/2175
- [x] `if let A(x) | B(x)`
- [x] `while let A(x) | B(x)`
- [ ] `let A(x) | B(x)`

### PR DESCRIPTION
**Pull request by Undin**
* Parse [or_patterns](https://github.com/rust-lang/rfcs/blob/master/text/2535-or-patterns.md)
* Provide name resolution and type inference
* Adapt existing code to new PSI. They should work in the same way as before in all cases supported before this PR
* Provide annotation that it's an experimental feature.
* Provide error about leading `|` in non-top level patterns

I didn't fix control flow graph and related inspections so they can work incorrectly in cases of new patterns

Fixes #3454
Fixes some points in #5032

### CODE REVIEW

**Comment by Undin** [comment]
CC @ortem

File: src/main/grammars/RustParser.bnf [hunk 1]
```diff
@@ -504,26 +504,32 @@ upper Function ::= async? const? unsafe? ExternAbi?
...
-      | PatTup
-      | PatSlice
-      | PatIdent
-      | PatMacro
-      | PatStruct
-      | PatTupleStruct
-      | (PatConst !('..' | '...' | '..='))
-      | PatRange
-      | PatBox {
+private TopPat ::= &'|' OrPat | !'!' Pat
```
Comment (vlad20012) [inline src/main/grammars/RustParser.bnf:507]:
```suggestion
private TopPat ::= &'|' OrPat | !'!' Pat { name = "pattern" }
```
Fixes chanes in `partial/exprs.txt` parsing test

Comment (vlad20012) [reply src/main/grammars/RustParser.bnf:507]: Also, why `!'!'` is needed? Let's add a test in `RsCompleteParsingTestCase` for this case

Comment (Undin) [reply src/main/grammars/RustParser.bnf:507]: > Also, why !'!' is needed?

:man_facepalming:
It should be `!'|'`. But it doesn't need anyway - just artifact of development. Thanks!

**Comment by Undin** [comment]
bors r=vlad20012

**Pull request merged**